---
title: "postHPC"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r}
suppressMessages(library(tidyverse))
suppressMessages(library(ArchR))
suppressMessages(library(GenomicFeatures))
suppressMessages(library(rtracklayer))
library(patchwork)
library(Seurat)
library(glue)
library(RColorBrewer)
library(ggrastr)
library(ggpubr)

source("common_scripts/graphs.R")
source("common_scripts/archrCleaning.R")
source("common_scripts/clusterAnnotation.R")
source("common_scripts/differentialExpression.R")
source("common_scripts/ncbiSearcher.R")

#https://stackoverflow.com/questions/23279904/modifying-an-r-package-function-for-current-r-session-assigninnamespace-not-beh
source("common_scripts/plotBrowserTrackCustom.R")

tmpFun <- get("plotBrowserTrack", envir = asNamespace("ArchR"))
environment(plotBrowserTrack2) <- environment(tmpFun)
attributes(plotBrowserTrack2) <- attributes(tmpFun)  # don't know if this is really needed
# assignInNamespace("plotBrowserTrack", plotBrowserTrack2, ns="ArchR")
```

```{r}
addArchRThreads(threads = 2)
```

```{r}
projInVitro <- loadArchRProject(path = "ND497_B_qcfiltTSS8/", showLogo = FALSE)

haystackInVitro <- addHaystackData(proj = projInVitro,
  haystackParentDir = "data/hiv-haystack",
  haystackSamples = c("ND497_B" = "ND497_B_v3"))

projInVitro <- haystackInVitro$newProj
```

```{r}
projChronic <- loadArchRProject(path = "C01C02_qcfiltTSS6/", showLogo = FALSE)

uniqChronicSamples <- unique(projChronic$Sample)
haystackSamplePairingChronic <- paste0(uniqChronicSamples, "_hxb2")
names(haystackSamplePairingChronic) <- uniqChronicSamples
  
haystackChronic <- addHaystackData(proj = projChronic,
  haystackParentDir = "data/hiv-haystack",
  haystackSamples = haystackSamplePairingChronic)

projChronic <- haystackChronic$newProj
```

```{r}
projART <- loadArchRProject(path = "A08A01_qcfiltTSS8/", showLogo = FALSE)

# add haystack data
uniqSamples <- unique(projART$Sample)
haystackSamplePairingART <- sapply(uniqSamples, function(i) {
  newSampleName <- ifelse(grepl("A01", i), paste0(i, "_v3WithEnv_filtered"), paste0(i, "_v3"))
  return(newSampleName)
})

haystackART <- addHaystackData(proj = projART,
  haystackParentDir = "data/hiv-haystack",
  haystackSamples = haystackSamplePairingART)

projART <- haystackART$newProj
```


```{r}
# load processed adt datasets
tsa_catalog <- readRDS("rds/tsa_catalog.rds")
load("rds/invitro_seuratMerged.RData")
load("rds/chronic_seuratMerged.RData")
load("rds/ART_seuratMerged.RData")
```


```{r}
# make final datasets that contains both modalities
inVitroIntersectCbc <- getIntersectBarcodes(adtInVitro, projInVitro)
chronicIntersectCbc <- getIntersectBarcodes(adtChronic, projChronic)
ARTIntersectCbc <- getIntersectBarcodes(adtART, projART)

projInvitro_matched <- projInVitro[inVitroIntersectCbc, ]
projInvitro_matched <- addUMAP(ArchRProj = projInvitro_matched,
  reducedDims = "IterativeLSI",
  name = "UMAP2",
  nNeighbors = 80,
  force = TRUE,
  minDist = 0.4)
projInvitro_matched <- addClusters(projInvitro_matched, method = "Seurat", name = "Clusters2", resolution = 0.8)
projInvitro_matched <- filterClusterByProp(projInvitro_matched, proportion = 0.5, cluster = "Clusters2")
adtInVitro_matched <- subset(adtInVitro, cells = projInvitro_matched$cellNames)

projChronic_matched <- projChronic[chronicIntersectCbc, ]
projChronic_matched <- addUMAP(ArchRProj = projChronic_matched,
  reducedDims = "Harmony",
  name = "UMAP2",
  nNeighbors = 80,
  force = TRUE,
  minDist = 0.3)

projChronic_matched <- addClusters(projChronic_matched,
  reducedDims = "Harmony",
  method = "Seurat",
  name = "Clusters2",
  force = TRUE,
  resolution = 0.5)
projChronic_matched <- filterClusterByProp(projChronic_matched, proportion = 0.5, cluster = "Clusters2")
adtChronic_matched <- subset(adtChronic, cells = projChronic_matched$cellNames)

projART_matched <- projART[ARTIntersectCbc, ]
projART_matched <- addUMAP(ArchRProj = projART_matched,
  reducedDims = "Harmony",
  name = "UMAP2",
  nNeighbors = 80,
  force = TRUE,
  minDist = 0.3)

projART_matched <- addClusters(projART_matched,
  reducedDims = "Harmony",
  method = "Seurat",
  name = "Clusters2",
  force = TRUE,
  maxClusters = NULL,
  resolution = 1)

projART_matched <- filterClusterByProp(projART_matched, 0.5, cluster = "Clusters2")
adtART_matched <- subset(adtART, cells = projART_matched$cellNames)
```


```{r}
# print out initial unlabeled umaps
plotUmap(projInvitro_matched, colorBy = "Clusters2", fn = "InVitro_umap_unlabeledCluster", embedding = "UMAP2")
plotUmap(projInvitro_matched,
  colorBy = "haystackOut",
  bringToTop = TRUE,
  embedding = "UMAP2",
  labelColors = FALSE,
  propInLegend = TRUE,
  fn = "InVitro_umap_haystackOut")

plotUmap(projChronic_matched, colorBy = "Clusters2", fn = "chronic_umap_unlabeledCluster", embedding = "UMAP2")
plotUmap(projChronic_matched,
  colorBy = "haystackOut",
  bringToTop = TRUE,
  labelColors = FALSE,
  embedding = "UMAP2",
  propInLegend = TRUE,
  fn = "chronic_umap_haystackOut")
plotUmap(projChronic_matched, colorBy = "individual", fn = "chronic_umap_indivdiual", embedding = "UMAP2", devices = "png")

plotUmap(projART_matched, colorBy = "Clusters2", fn = "art_umap_unlabeledCluster", embedding = "UMAP2")
plotUmap(projART_matched,
  colorBy = "haystackOut",
  bringToTop = TRUE,
  labelColors = FALSE,
  embedding = "UMAP2",
  fn = "art_umap_haystackOut")
plotUmap(projART_matched, colorBy = "individual", fn = "art_umap_indivdiual", embedding = "UMAP2", devices = "png")
```


```{r}
# select features that aren't similar to isotype...not sure if this will really be necessary
inVitroIsoSimilar <- returnSimilarToIsotype(tsa_catalog, isoComparisonsInvitro, 0.05, 4)
chronicIsoSimilar <- returnSimilarToIsotype(tsa_catalog, isoComparisonsChronic, 0.05, 4)
ARTIsoSimilar <- returnSimilarToIsotype(tsa_catalog, isoComparisonsART, 0.05, 4)

inVitroAdtToUse <- returnADTNonSimilar(adtInVitro_matched, tsa_catalog, inVitroIsoSimilar)
chronicAdtToUse <- returnADTNonSimilar(adtChronic_matched, tsa_catalog, chronicIsoSimilar)
ARTAdtToUse <- returnADTNonSimilar(adtART_matched, tsa_catalog, ARTIsoSimilar)

adtInVitro_matched$atacClusters <- projInvitro_matched$Clusters2
adtChronic_matched$atacClusters <- projChronic_matched$Clusters2
adtART_matched$atacClusters <- projART_matched$Clusters2

projInvitro_matched <- addImputeWeights(projInvitro_matched)
getBaseATACPanel(projInvitro_matched, "invitro_imputePanel", embedding = "UMAP2")
getBasePanelAndMarkers(seu = adtInVitro_matched, tsa_catalog = tsa_catalog, featuresToUse = inVitroAdtToUse,
  findMarkerMethod = "wilcox",
  plotFn = "InVitro_basepanel", tsvFn = "outs/tsv/InVitro_initClusterMarkers.tsv")

projChronic_matched <- addImputeWeights(projChronic_matched)
getBaseATACPanel(projChronic_matched, "chronic_imputePanel", embedding = "UMAP2")
getBasePanelAndMarkers(seu = adtChronic_matched, tsa_catalog = tsa_catalog, featuresToUse = chronicAdtToUse,
  findMarkerMethod = "wilcox",
  plotFn = "chronic_basepanel", tsvFn = "outs/tsv/chronic_initClusterMarkers.tsv")

projART_matched <- addImputeWeights(projART_matched)
getBaseATACPanel(projART_matched, "art_imputePanel", embedding = "UMAP2")
getBasePanelAndMarkers(seu = adtART_matched, tsa_catalog, featuresToUse = ARTAdtToUse,
  findMarkerMethod = "wilcox",
  plotFn = "art_basepanel", tsvFn = "outs/tsv/art_initClusterMarkers.tsv")
```


```{r}
projInvitro_matched <- assignManualAnnotation(archrProj = projInvitro_matched,
  cluster = "Clusters2",
  annotFn = "manualClusterAnnotations/invitro.csv")

adtInVitro_matched$manualClusterAnnot <- projInvitro_matched$manualClusterAnnot
adtInVitro_matched$haystackOut <- projInvitro_matched$haystackOut

plotUmap(projInvitro_matched, colorBy = "manualClusterAnnot", fn = "InVitro_umap_labeledCluster", embedding = "UMAP2")
plotUmap(projInvitro_matched, colorBy = "manualClusterAnnot", fn = "InVitro_umap_labeledCluster_noPlotLabels", embedding = "UMAP2", labelColors = FALSE, propInLegend = TRUE)

plotDiscreteLollipop(projInvitro_matched, "inVitro_discreteAbsolute_matched",
  cluster = "manualClusterAnnot",
  graphType = "absolute")

plotDiscreteLollipop(projInvitro_matched, "inVitro_discreteHivOnly_matched",
  cluster = "manualClusterAnnot",
  graphType = "hivOnly")
```


```{r}
projChronic_matched <- assignManualAnnotation(archrProj = projChronic_matched,
  cluster = "Clusters2",
  annotFn = "manualClusterAnnotations/chronic.csv")

adtChronic_matched$manualClusterAnnot <- projChronic_matched$manualClusterAnnot
adtChronic_matched$haystackOut <- projChronic_matched$haystackOut

plotUmap(projChronic_matched, colorBy = "manualClusterAnnot",
  fn = "chronic_umap_labeledCluster",
  embedding = "UMAP2")
plotUmap(projChronic_matched, colorBy = "manualClusterAnnot",
  fn = "chronic_umap_labeledCluster_noPlotLabels",
  embedding = "UMAP2",
  propInLegend = TRUE,
  labelColors = FALSE)

plotDiscreteLollipop(projChronic_matched, "chronic_discreteAbsolute_matched",
  cluster = "manualClusterAnnot",
  graphType = "absolute")

plotDiscreteLollipop(projChronic_matched, "chronic_discreteHivOnly_matched",
  cluster = "manualClusterAnnot",
  graphType = "hivOnly")
```


```{r}
projART_matched <- assignManualAnnotation(archrProj = projART_matched,
  cluster = "Clusters2",
  annotFn = "manualClusterAnnotations/art.csv")

adtART_matched$manualClusterAnnot <- projART_matched$manualClusterAnnot
adtART_matched$haystackOut <- projART_matched$haystackOut

plotUmap(projART_matched, colorBy = "manualClusterAnnot", fn = "art_umap_labeledCluster", embedding = "UMAP2")
plotUmap(projART_matched, colorBy = "manualClusterAnnot", fn = "art_umap_labeledCluster_noPlotLabels", labelColors = FALSE, embedding = "UMAP2")

plotDiscreteLollipop(projART_matched, "art_discreteAbsolute_matched",
  cluster = "manualClusterAnnot",
  gheight = 4,
  graphType = "absolute")

plotDiscreteLollipop(projART_matched, "art_discreteHivOnly_matched",
  cluster = "manualClusterAnnot",
  gheight = 4,
  graphType = "hivOnly")
```

### BEGIN DIFFERENTIAL ANALYSIS

## Invitro differential analysis
```{r}
invitro_tcells_check <- grepl("^CD4", adtInVitro_matched$manualClusterAnnot)
invitro_tcells <- adtInVitro_matched$haystackOut[invitro_tcells_check]
invitro_markers <- findHIVDifferentialMarkers(seu = adtInVitro_matched,
  cellsPos = names(invitro_tcells[invitro_tcells]),
  cellsNeg = names(invitro_tcells[!invitro_tcells]),
  featuresToUse = inVitroAdtToUse,
  tsa_catalog = tsa_catalog)

makeDifferentialLollipop(invitro_markers, "invitro_lollipop")

invitro_activeLaterCells <- adtInVitro_matched$haystackOut[grepl("CD4 Activated", adtInVitro_matched$manualClusterAnnot) & invitro_tcells_check]
invitro_activeLate_markers <- findHIVDifferentialMarkers(
  seu = adtInVitro_matched,
  tsa_catalog = tsa_catalog,
  cellsPos = names(invitro_activeLaterCells[invitro_activeLaterCells]),
  cellsNeg = names(invitro_activeLaterCells[!invitro_activeLaterCells]),
  featuresToUse = inVitroAdtToUse)

makeDifferentialLollipop(invitro_activeLate_markers, "invitro_lollipop_activeLate")

invitro_activeLate_markers_forVln <- invitro_activeLate_markers %>%
  group_by(Status) %>%
  arrange(desc(abs(piScore)), .by_group = TRUE) %>%
  top_n(10, abs(piScore))

plotVlnEnhanced(adtInVitro_matched, cells = names(invitro_activeLaterCells),
  feats = invitro_activeLate_markers_forVln$gene,
  xVar = "haystackOut",
  showAggregate = FALSE,
  separator = invitro_activeLate_markers_forVln$Status,
  titles = invitro_activeLate_markers_forVln$cleanName,
  fn = "invitro_activated_vln_top10")

invitro_earlyCells <- adtInVitro_matched$haystackOut[grepl("(Tcm|early)", adtInVitro_matched$manualClusterAnnot) & invitro_tcells_check]
invitro_early_markers <- findHIVDifferentialMarkers(
  seu = adtInVitro_matched,
  tsa_catalog = tsa_catalog,
  cellsPos = names(invitro_earlyCells[invitro_earlyCells]),
  cellsNeg = names(invitro_earlyCells[!invitro_earlyCells]),
  featuresToUse = inVitroAdtToUse)

makeDifferentialLollipop(invitro_early_markers, "invitro_lollipop_early")


invitroGcPeakFn <- paste0(getOutputDirectory(projInvitro_matched), "_gcPeakCalled")
if (!dir.exists(invitroGcPeakFn)) {
  invitro_renamingDF <- getCellColData(projInvitro_matched, select = c("haystackOut", "manualClusterAnnot")) %>%
    as.data.frame(.) %>%
    mutate(cellNames = projInvitro_matched$cellNames) %>%
    mutate(condensedCluster = case_when(
      grepl("Activated", manualClusterAnnot) ~ "Activated",
      grepl("Tcm", manualClusterAnnot) ~ "Tcm",
      grepl("Naive", manualClusterAnnot) ~ "Naive_earlyMemory",
      grepl("Treg", manualClusterAnnot) ~ "Treg",
      TRUE ~ "Other")) %>%
    mutate(condensedHIVCluster = paste0(condensedCluster, "_", haystackOut))

  projInvitro_matched <- addCellColData(ArchRProj = projInvitro_matched,
    data = invitro_renamingDF$condensedHIVCluster,
    cells = invitro_renamingDF$cellNames,
    name = "condensedHivCluster",
    force = TRUE)
  
  projInvitro_matched <- addGroupCoverages(ArchRProj = projInvitro_matched,
    groupBy = "condensedHivCluster",
    minReplicates = 3,
    force = TRUE)
  
  projInvitro_matched <- addReproduciblePeakSet(
    ArchRProj = projInvitro_matched,
    groupBy = "condensedHivCluster", 
    pathToMacs2 = "/Users/vince/anaconda3/envs/asapseq/bin/macs2",
    force = TRUE
  )
  
  projInvitro_matched <- addPeakMatrix(projInvitro_matched, force = TRUE)
  projInvitro_matched <- addMotifAnnotations(ArchRProj = projInvitro_matched,
    motifSet = "cisbp",
    name = "Motif",
    force = TRUE)
  
  projInvitro_matched <- addBgdPeaks(projInvitro_matched)
  projInvitro_matched <- addDeviationsMatrix(
    ArchRProj = projInvitro_matched, 
    peakAnnotation = "Motif",
    force = TRUE
  )

  projInvitro_matched_gcPeak <- saveArchRProject(projInvitro_matched,
    outputDirectory = invitroGcPeakFn)
} else {
  projInvitro_matched_gcPeak <- loadArchRProject(invitroGcPeakFn, force = TRUE, showLogo = FALSE)
}

invitro_markerTest_Activated <- getMarkerFeatures(
  ArchRProj = projInvitro_matched_gcPeak, 
  useMatrix = "PeakMatrix",
  groupBy = "condensedHivCluster",
  testMethod = "wilcoxon",
  bias = c("TSSEnrichment", "log10(nFrags)"),
  useGroups = "Activated_TRUE",
  bgdGroups = "Activated_FALSE",
  maxCells = 1000,
  verbose = FALSE
)

invitro_markerTest_activatedPeaks <- getMarkers(invitro_markerTest_Activated, cutOff = "FDR < 0.05", returnGR = TRUE)
invitro_markerTest_activatedPeaks <- invitro_markerTest_activatedPeaks$Activated_TRUE
invitro_markerTest_activatedPeaks <- hiAnnotator::getNearestFeature(invitro_markerTest_activatedPeaks,
  promoters(getGeneAnnotation(projInvitro_matched_gcPeak)[[1]], upstream = 0, downstream = 1), colnam = "nearestTSS",
  feature.colnam = "symbol")

invitro_markerTest_activatedPeaks <- hiAnnotator::getSitesInFeature(invitro_markerTest_activatedPeaks,
  getGeneAnnotation(projInvitro_matched_gcPeak)[[1]], colnam = "inGene",
  feature.colnam = "symbol")

invitro_markerTest_activatedPeaks_topPeaks <- data.frame(invitro_markerTest_activatedPeaks) %>%
  mutate(direction = ifelse(Log2FC > 0, "HIV+", "HIV-")) %>%
  mutate(piScore = Log2FC * -log10(FDR)) %>%
  group_by(direction) %>%
  slice_max(order_by = abs(piScore), n = 15) %>%
  arrange(abs(piScore), .by_group = TRUE) %>%
  ungroup() %>%
  mutate(symbol2 = case_when(
    inGene != "FALSE" & str_detect(as.character(inGene), nearestTSS) ~ glue("{nearestTSS}* ({nearestTSSDist})"),
    inGene != "FALSE" ~ glue("{inGene}*; {nearestTSS} ({nearestTSSDist})"),
    TRUE ~ glue("{nearestTSS} ({nearestTSSDist})"))) %>%
  mutate(symbol2 = str_wrap(symbol2, width = 35, exdent = 2)) %>%
  mutate(symbol2 = factor(symbol2, levels = symbol2))

iv_topPeaks_g <- ggplot(invitro_markerTest_activatedPeaks_topPeaks, aes(x = -log10(FDR), y = symbol2)) +
  geom_segment(aes(yend = symbol2, x = 0, xend = -log10(FDR)), linetype = "dashed", color = "#CCCCCC") +
  geom_point(aes(color = Log2FC)) +
  scale_color_distiller(palette = "RdBu", direction = -1) +
  theme_classic() +
  theme(panel.background = element_rect(fill = NULL, colour = NULL),
    plot.margin = margin(0,0,0,0),
    strip.background = element_blank(),
    strip.text = element_text(size = BASEPTFONTSIZE, margin = margin(0, 0, 0, 0)),
    text = element_text(family = "Arial", size = BASEPTFONTSIZE),
    axis.text = element_text(size = 6, color = "#000000"),
    axis.title.y = element_blank(),
    legend.text = element_text(size = 6),
    legend.title = element_text(size = BASEPTFONTSIZE),
    legend.box.margin = margin(0,0,0,-10),
    legend.direction = "vertical",
    legend.position = "right",
    legend.key.height = unit(0.5, "line"),
    legend.key.width = unit(0.5, "line")) +
  labs(color = "log2(FC)") +
  scale_x_continuous(limits = c(0,NA)) +
  coord_cartesian(clip = "off") +
  scale_size_area(max_size = 4) +
  facet_wrap(~ direction, scales = "free")

savePlot(iv_topPeaks_g, fn = "invitro_markerTest_activated_topPeaks", devices = c("png", "rds"), gheight = 3.5, gwidth = 5)

plotVolcanoFromGetMarkerFeatures(invitro_markerTest_Activated,
  fn = "invitro_peaks_volcano_activatedHIV")

motifsUp <- peakAnnoEnrichment(
  seMarker = invitro_markerTest_Activated,
  ArchRProj = projInvitro_matched_gcPeak,
  peakAnnotation = "Motif",
  cutOff = "FDR < 0.05 & Log2FC > 0")

motifsDown <- peakAnnoEnrichment(
  seMarker = invitro_markerTest_Activated,
  ArchRProj = projInvitro_matched_gcPeak,
  peakAnnotation = "Motif",
  cutOff = "FDR < 0.05 & Log2FC < 0")

plotMotifRank(motifsDown, "invitro_motifsUp_activatedHIVneg", color = HIVNEGCOLOR)
plotMotifRank(motifsUp, "invitro_motifsUp_activatedHIVpos", color = HIVPOSCOLOR)


iv_ccr5_track <- plotBrowserTrack2(projInvitro_matched_gcPeak,
  geneSymbol = "CCR5",
  useGroups = c("Activated_FALSE", "Activated_TRUE"),
  groupBy = "condensedHivCluster",
  upstream = 10000, downstream = 10000,
  scCellsMax = 750,
  plotSummary = c("bulkTrack", "scTracK", "geneTrack"),
  pal = c(HIVNEGCOLOR, HIVPOSCOLOR), borderWidth = 0, sizes = c(3,2,1))$CCR5

cleanUpTrackAndSave(iv_ccr5_track, fn = "invitro_activated_ccr5")

plotVolcanoFromGetMarkerFeatures(invitro_markerTest_tcm_activated,
  fn = "invitro_peaks_volcano_tcmActivatedHIV")

invitro_markerTest_tcm <- getMarkerFeatures(
  ArchRProj = projInvitro_matched_gcPeak, 
  useMatrix = "PeakMatrix",
  groupBy = "condensedHivCluster",
  testMethod = "wilcoxon",
  bias = c("TSSEnrichment", "log10(nFrags)"),
  useGroups = "Tcm_TRUE",
  bgdGroups = "Tcm_FALSE",
  maxCells = 1000,
  verbose = FALSE
)

plotVolcanoFromGetMarkerFeatures(invitro_markerTest_tcm,
  fn = "invitro_peaks_volcano_tcmHIV")

# chromVAR analysis
invitro_chromvar_markerTest <- getMarkerFeatures(
  ArchRProj = projInvitro_matched_gcPeak,
  useMatrix = "MotifMatrix",
  groupBy = "condensedHivCluster",
  testMethod = "wilcoxon",
  bias = c("TSSEnrichment", "log10(nFrags)"),
  useGroups = "Activated_TRUE",
  bgdGroups = "Activated_FALSE",
  maxCells = 1000,
  verbose = FALSE,
  useSeqnames = "deviations"
)

plotVolcanoFromGetMarkerFeatures(invitro_chromvar_markerTest, chromVARmode = TRUE,
  fn = "invitro_volcano_motifs_chromVAR")

plotMotifDot(invitro_chromvar_markerTest, "invitro_chromVAR_motifsUp_activatedHIVneg", direction = "negative")
plotMotifDot(invitro_chromvar_markerTest, "invitro_chromVAR_motifsUp_activatedHIVpos", direction = "positive")

invitro_chromvar_markerTest_tcm <- getMarkerFeatures(
  ArchRProj = projInvitro_matched_gcPeak,
  useMatrix = "MotifMatrix",
  groupBy = "condensedHivCluster",
  testMethod = "wilcoxon",
  bias = c("TSSEnrichment", "log10(nFrags)"),
  useGroups = "Tcm_TRUE",
  bgdGroups = "Tcm_FALSE",
  maxCells = 1000,
  verbose = FALSE,
  useSeqnames = "deviations"
)

plotVolcanoFromGetMarkerFeatures(invitro_chromvar_markerTest_tcm, chromVARmode = TRUE,
  fn = "invitro_volcano_motifs_chromVAR_tcm")
```


## Chronic differential analysis
```{r}
chronic_tcells_check <- !(grepl("(B|APC|CD8)", adtChronic_matched$manualClusterAnnot))
chronic_tcells <- adtChronic_matched$haystackOut[chronic_tcells_check]
chronic_markers <- findHIVDifferentialMarkers(seu = adtChronic_matched,
  cellsPos = names(chronic_tcells[chronic_tcells]),
  cellsNeg = names(chronic_tcells[!chronic_tcells]),
  featuresToUse = chronicAdtToUse,
  findMarkerMethod = "DESeq2",
  tsa_catalog = tsa_catalog)

plotVlnEnhanced(adtChronic_matched, cells = names(chronic_tcells),
  feats = chronic_markers$gene,
  separator = chronic_markers$Status,
  titles = chronic_markers$cleanName,
  fn = "chronic_tcells_vln")

makeDifferentialLollipop(chronic_markers, "chronic_lollipop")
 
chronic_tfh_cells <- adtChronic_matched$haystackOut[grepl("(CD4 Tfh)", adtChronic_matched$manualClusterAnnot)]
chronic_tfh_markers <- findHIVDifferentialMarkers(
  seu = adtChronic_matched,
  tsa_catalog = tsa_catalog,
  cellsPos = names(chronic_tfh_cells[chronic_tfh_cells]),
  cellsNeg = names(chronic_tfh_cells[!chronic_tfh_cells]),
  findMarkerMethod = "DESeq2",
  featuresToUse = chronicAdtToUse)

makeDifferentialLollipop(chronic_tfh_markers, "chronic_lollipop_tfh")

plotVlnEnhanced(adtChronic_matched, cells = names(chronic_tfh_cells),
  feats = chronic_tfh_markers$gene,
  separator = chronic_tfh_markers$Status,
  titles = chronic_tfh_markers$cleanName,
  fn = "chronic_tfh_vln")

chronicGcPeakFn <- paste0(getOutputDirectory(projChronic_matched), "_gcPeakCalled")
if (!dir.exists(chronicGcPeakFn)) {
  chronic_renamingDF <- getCellColData(projChronic_matched, select = c("haystackOut", "manualClusterAnnot")) %>%
    as.data.frame(.) %>%
    mutate(cellNames = projChronic_matched$cellNames) %>%
    mutate(condensedCluster = case_when(
      grepl("^CD4", manualClusterAnnot) ~ "CD4",
      TRUE ~ "Other")) %>%
    mutate(condensedHIVCluster = paste0(condensedCluster, "_", haystackOut))
  
  projChronic_matched <- addCellColData(ArchRProj = projChronic_matched,
    data = chronic_renamingDF$condensedHIVCluster,
    cells = chronic_renamingDF$cellNames,
    name = "condensedHivCluster",
    force = TRUE)
  
  projChronic_matched <- addGroupCoverages(ArchRProj = projChronic_matched,
    groupBy = "condensedHivCluster",
    minReplicates = 3,
    minCells = 15,
    sampleRatio = 0.95,
    force = TRUE)
  
  projChronic_matched <- addReproduciblePeakSet(
    ArchRProj = projChronic_matched,
    groupBy = "condensedHivCluster", 
    pathToMacs2 = "/Users/vince/anaconda3/envs/asapseq/bin/macs2",
    force = TRUE)
  
  projChronic_matched <- addPeakMatrix(projChronic_matched, force = TRUE)
  
  projChronic_matched_gcPeak <- saveArchRProject(projChronic_matched,
    outputDirectory = chronicGcPeakFn)
} else {
  projChronic_matched_gcPeak <- loadArchRProject(chronicGcPeakFn, force = TRUE, showLogo = FALSE)
}

chronic_markerTest <- getMarkerFeatures(
  ArchRProj = projChronic_matched_gcPeak, 
  useMatrix = "PeakMatrix",
  groupBy = "condensedHivCluster",
  testMethod = "wilcoxon",
  bias = c("TSSEnrichment", "log10(nFrags)"),
  useGroups = "CD4_TRUE",
  bgdGroups = "CD4_FALSE",
  maxCells = 1000,
  verbose = FALSE
)
plotVolcanoFromGetMarkerFeatures(chronic_markerTest,
  fn = "chronic_peaks_volcano")
```


## chronic chromVAR analysis
```{r}
if("Motif" %ni% names(projChronic_matched_gcPeak@peakAnnotation)){
    projChronic_matched_gcPeak <- addMotifAnnotations(ArchRProj = projChronic_matched_gcPeak, motifSet = "cisbp", name = "Motif")
}

projChronic_matched_gcPeak <- addBgdPeaks(projChronic_matched_gcPeak)

projChronic_matched_gcPeak <- addDeviationsMatrix(
  ArchRProj = projChronic_matched_gcPeak, 
  peakAnnotation = "Motif",
  force = TRUE
)

chronic_markerTest <- getMarkerFeatures(
  ArchRProj = projChronic_matched_gcPeak,
  useMatrix = "MotifMatrix",
  groupBy = "condensedHivCluster",
  testMethod = "wilcoxon",
  bias = c("TSSEnrichment", "log10(nFrags)"),
  useGroups = "CD4_TRUE",
  bgdGroups = "CD4_FALSE",
  maxCells = 1000,
  verbose = FALSE,
  useSeqnames = "deviations"
)

plotVolcanoFromGetMarkerFeatures(chronic_markerTest, chromVARmode = TRUE,
  fn = "chronic_volcano_motifs_chromVAR")

plotMotifDot(chronic_markerTest, "chronic_chromVAR_motifsUp_HIVneg", direction = "negative")
plotMotifDot(chronic_markerTest, "chronic_chromVAR_motifsUp_HIVpos", direction = "positive")
```


## ART differential analysis
```{r}
art_tcells_check <- (grepl("CD4", adtART_matched$manualClusterAnnot))
art_tcells <- adtART_matched$haystackOut[art_tcells_check]
art_markers <- findHIVDifferentialMarkers(seu = adtART_matched,
  cellsPos = names(art_tcells[art_tcells]),
  cellsNeg = names(art_tcells[!art_tcells]),
  featuresToUse = ARTAdtToUse,
  findMarkerMethod = "DESeq2",
  tsa_catalog = tsa_catalog)

makeDifferentialLollipop(art_markers, "art_lollipop")

art_renamingDF <- getCellColData(projART_matched, select = c("haystackOut", "manualClusterAnnot")) %>%
  as.data.frame(.) %>%
  mutate(cellNames = projART_matched$cellNames) %>%
  mutate(condensedCluster = case_when(
    grepl("CD4", manualClusterAnnot) ~ "T cell",
    TRUE ~ "Other")) %>%
  mutate(condensedHIVCluster = paste0(condensedCluster, "_", haystackOut))

projART_matched <- addCellColData(ArchRProj = projART_matched,
  data = art_renamingDF$condensedHIVCluster,
  cells = art_renamingDF$cellNames,
  name = "condensedHivCluster",
  force = TRUE)

projART_matched <- addGroupCoverages(ArchRProj = projART_matched,
  groupBy = "condensedHivCluster",
  minReplicates = 2,
  minCells = 15,
  sampleRatio = 0.95,
  force = TRUE)

projART_matched <- addReproduciblePeakSet(
  ArchRProj = projART_matched,
  groupBy = "condensedHivCluster", 
  pathToMacs2 = "/Users/vince/anaconda3/envs/asapseq/bin/macs2",
  force = TRUE)

projART_matched <- addPeakMatrix(projART_matched, force = TRUE)

if("Motif" %ni% names(projART_matched@peakAnnotation)){
    projART_matched <- addMotifAnnotations(ArchRProj = projART_matched, motifSet = "cisbp", name = "Motif")
}

projART_matched <- addBgdPeaks(projART_matched)

projART_matched <- addDeviationsMatrix(
  ArchRProj = projART_matched, 
  peakAnnotation = "Motif",
  force = TRUE
)

art_markerTest <- getMarkerFeatures(
  ArchRProj = projART_matched, 
  useMatrix = "MotifMatrix",
  groupBy = "condensedHivCluster",
  testMethod = "wilcoxon",
  bias = c("TSSEnrichment", "log10(nFrags)"),
  useGroups = "T cell_TRUE",
  bgdGroups = "T cell_FALSE",
  maxCells = 1000,
  verbose = FALSE,
  useSeqnames = "deviations"
)

plotVolcanoFromGetMarkerFeatures(art_markerTest, chromVARmode = TRUE,
  fn = "art_volcano_motifs_chromVAR")
```

# qc metrics
```{r}
invitro_qc_nfrags <- plotArchRQCData(projInvitro_matched, fn = "invitro_qc_haystack_frags",
  ycol = "log10(nFrags)", ylbl = "log10(unique fragments)")
invitro_qc_nfrags %>%
  group_by(xaxis) %>%
  summarize(avg = mean(yaxis), 
    sd = stats::sd(yaxis))
  
invitro_qc_tss <- plotArchRQCData(projInvitro_matched, fn = "invitro_qc_haystack_TSS", ycol = "TSSEnrichment")
invitro_qc_tss %>%
  group_by(xaxis) %>%
  summarize(avg = mean(yaxis), 
    sd = stats::sd(yaxis))

plotUpsetCBC(atacCBC = projInVitro$cellNames,
  hivCBC = unique(haystackInVitro$allViralFrags$newCbc),
  adtCBC = Cells(adtInVitro),
  fn = "invitro_upset_cbc")

chronic_qc_nfrags <- plotArchRQCData(projChronic_matched, fn = "chronic_qc_haystack_frags",
  ycol = "log10(nFrags)", ylbl = "log10(unique fragments)")
chronic_qc_nfrags %>%
  group_by(xaxis) %>%
  summarize(avg = mean(yaxis), 
    sd = stats::sd(yaxis))
  

chronic_qc_tss <- plotArchRQCData(projChronic_matched, fn = "chronic_qc_haystack_TSS", ycol = "TSSEnrichment")
chronic_qc_tss %>%
  group_by(xaxis) %>%
  summarize(avg = mean(yaxis), 
    sd = stats::sd(yaxis))

plotUpsetCBC(atacCBC = projChronic$cellNames,
  hivCBC = unique(haystackChronic$allViralFrags$newCbc),
  adtCBC = Cells(adtChronic),
  fn = "chronic_upset_cbc")
```

# LTR vs internal coverage
```{r}
calculateLTRVersusInternal <- function(frags, ltr5End, ltr3Start, ltr3End) {
  df <- frags %>%
    mutate(ltrOnly = (startBp <= ltr5End & endBp <= ltr5End) | (startBp >= ltr3Start & endBp <= ltr3End)) %>%
    mutate(internalOnly = startBp > ltr5End & endBp < ltr3Start) %>%
    mutate(both = !ltrOnly & !internalOnly) %>%
    summarize(LTRonly = sum(ltrOnly),
      internalOnly = sum(internalOnly),
      both = sum(both)) %>%
    pivot_longer(everything(), names_to = "metric", values_to = "n") %>%
    mutate(prop = round(n / sum(n), digits = 2))
  
  return(df)
}

sumaGeneAnnots <- read.csv("viralGenomeAnnotations/suma.csv") %>%
  mutate(annotation = factor(annotation, levels = rev(unique(annotation)))) %>%
  mutate(annotation2 = ifelse(annotation == "LTR", paste0(annotation, row_number()), annotation))

inVitroFrags_matched <- haystackInVitro$filteredviralFrags[haystackInVitro$filteredviralFrags$newCbc %in% projInvitro_matched$cellNames, ]

plotFragMultiGraph(frags = inVitroFrags_matched, coverage = rep(0, 9726), geneAnnots = sumaGeneAnnots, fn = "invitro_frags")
calculateLTRVersusInternal(inVitroFrags_matched, 632, 9093, 9725)

# inVitroFrags_matched %>%
#   group_by(newCbc) %>%
#   distinct(readname) %>%
#   dplyr::summarize(totalUniq = n()) %>%
#   left_join(invitro_qc_nfrags, by = c("newCbc" = "cell")) %>%
#   ggplot(aes(x = totalUniq, y = 10 ** yaxis)) +
#   geom_point() +
#   stat_smooth(method = "lm", formula = y ~ x)

hxb2GeneAnnots <- read.csv("viralGenomeAnnotations/hxb2.csv") %>%
  mutate(annotation = factor(annotation, levels = rev(unique(annotation)))) %>%
  mutate(annotation2 = ifelse(annotation == "LTR", paste0(annotation, row_number()), annotation))

chronicFrags_matched <- haystackChronic$filteredviralFrags[haystackChronic$filteredviralFrags$newCbc %in% projChronic_matched$cellNames, ] %>%
  mutate(individual = ifelse(grepl("C01", sample), "C01", "C02"))

plotFragMultiGraph(frags = chronicFrags_matched, coverage = rep(0, 9719), geneAnnots = hxb2GeneAnnots, separateByIndividual = TRUE, fn = "chronic_frags", gheight = 6)
calculateLTRVersusInternal(chronicFrags_matched, 633, 9085, 9718)

artFrags_matched <- haystackART$filteredviralFrags[haystackART$filteredviralFrags$newCbc %in% projART_matched$cellNames, ]
artVirusGeneAnnots <- downloadTables(unique(artFrags_matched$seqname))

checkFragmentOverlap <- function(q1, q2, s1, s2) {
  if (q1 <= s2 & q1 >= s1) {
    return(TRUE)
  } else if (q2 >= s1 & q2 <= s2) {
    return(TRUE)
  } else if (s2 >= q1 & s2 <= q2) {
    return(TRUE)
  } else if (s1 >= q1 & s1 <= q2) {
    return(TRUE)
  }
  return(FALSE)
}

sapply(seq_along(artFrags_matched$sample), function(i) {
  seqname <- artFrags_matched[i, "seqname"]
  seqnameAnnots <- artVirusGeneAnnots[[seqname]]
  
  if (is.null(seqnameAnnots)) {
    return(NULL)
  }
  
  q1 <- artFrags_matched[i, "startBp"]
  q2 <- artFrags_matched[i, "endBp"]
  
  matches <- sapply(seq_along(seqnameAnnots$annotation), function(i) {
    s1 <- seqnameAnnots$startPos[i]
    s2 <- seqnameAnnots$endPos[i]
    
    return(checkFragmentOverlap(q1, q2, s1, s2))
  })
  
  return(seqnameAnnots$annotation[matches])
})
# TODO need to add...LTR sequences for some sequences since they're not annotated.
```


# get numbers for manuscript
```{r}
# iv total cells
length(projInvitro_matched$cellNames)

# iv count HIV+
table(projInvitro_matched$haystackOut)

# iv percent HIV+
round(table(projInvitro_matched$haystackOut) / length(projInvitro_matched$haystackOut) * 100, digits = 1)

# iv percent annot HIV+ were activated/effector
round(length(projInvitro_matched_gcPeak$manualClusterAnnot[projInvitro_matched_gcPeak$haystackOut &
    grepl("Activated", projInvitro_matched_gcPeak$manualClusterAnnot)]) / sum(projInvitro_matched_gcPeak$haystackOut) * 100,
  digits = 1)

# iv percent annot HIV+ were CD4 that were not activated/effector
round(length(projInvitro_matched_gcPeak$manualClusterAnnot[projInvitro_matched_gcPeak$haystackOut &
    grepl("CD4 (Naive|T)", projInvitro_matched_gcPeak$manualClusterAnnot)]) / sum(projInvitro_matched_gcPeak$haystackOut) * 100,
  digits = 1)

# iv percent annot HIV+ were mono
round(length(projInvitro_matched_gcPeak$manualClusterAnnot[projInvitro_matched_gcPeak$haystackOut &
    grepl("Monocyte", projInvitro_matched_gcPeak$manualClusterAnnot)]) / sum(projInvitro_matched_gcPeak$haystackOut) * 100,
  digits = 1)


### CHRONIC
# chronic percent total cells
table(projChronic_matched$individual)

# chronic count HIV+
table(projChronic_matched$individual, projChronic_matched$haystackOut)

# chronic percent HIV+
tmp <- table(projChronic_matched$individual, projChronic_matched$haystackOut) 
round(tmp / rowSums(tmp) * 100, digits = 2)

# chronic total
round(table(projChronic_matched$haystackOut) / length(projChronic_matched$cellNames) * 100, digits = 2)

# chronic percent is memory cd4
length(projChronic_matched$manualClusterAnnot[grepl("^CD4", projChronic_matched$manualClusterAnnot)]) / length(projChronic_matched$cellNames) * 100

# percent of all clusters
table(projChronic_matched$manualClusterAnnot) / length(projChronic_matched$cellNames) * 100

# percent HIV+ in Trm or Tcm
length(projChronic_matched$manualClusterAnnot[grepl("(Trm|Tcm)", projChronic_matched$manualClusterAnnot) & projChronic_matched$haystackOut]) / length(projChronic_matched$manualClusterAnnot[projChronic_matched$haystackOut])


### ART
# art percent total cells
table(projART_matched$individual)

# art count HIV+
tmp <- table(projART_matched$individual, projART_matched$haystackOut)

# art percent HIV+
round(tmp / rowSums(tmp) * 100, digits = 2)

# art total
length(projART_matched$cellNames)

# art total
table(projART_matched$haystackOut)

# art percent total
round(table(projART_matched$haystackOut) / length(projART_matched$cellNames) * 100, digits = 2)
```

