---
title: "postHPC"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r}
library(tidyverse)
suppressMessages(library(ArchR))
suppressMessages(library(GenomicFeatures))
suppressMessages(library(rtracklayer))
library(patchwork)
library(pheatmap)
library(UpSetR)
library(Seurat)
library(glue)

source("common_scripts/graphs.R")
source("common_scripts/archrCleaning.R")
source("common_scripts/clusterAnnotation.R")
```

```{r}
addArchRThreads(threads = 2)
```

```{r}
projInVitro <- loadArchRProject(path = "ND497_B_qcfiltTSS8/", showLogo = FALSE)

haystackInVitro <- addHaystackData(proj = projInVitro,
  haystackParentDir = "data/hiv-haystack",
  haystackSamples = c("ND497_B" = "ND497_B_v3"))

projInVitro <- haystackInVitro$newProj
```

```{r}
projChronic <- loadArchRProject(path = "C01C02_qcfiltTSS6/", showLogo = FALSE)

uniqChronicSamples <- unique(projChronic$Sample)
haystackSamplePairingChronic <- paste0(uniqChronicSamples, "_hxb2")
names(haystackSamplePairingChronic) <- uniqChronicSamples
  
haystackChronic <- addHaystackData(proj = projChronic,
  haystackParentDir = "data/hiv-haystack",
  haystackSamples = haystackSamplePairingChronic)

projChronic <- haystackChronic$newProj

projChronic <- addUMAP(ArchRProj = projChronic,
  reducedDims = "Harmony",
  name = "UMAP2",
  nNeighbors = 40,
  minDist = 0.4)
```

```{r}
projART <- loadArchRProject(path = "A08A01_qcfiltTSS8/", showLogo = FALSE)

# add haystack data
uniqSamples <- unique(projART$Sample)
haystackSamplePairingART <- sapply(uniqSamples, function(i) {
  newSampleName <- ifelse(grepl("A01", i), paste0(i, "_v3WithEnv"), paste0(i, "_v3"))
  return(newSampleName)
})

haystackART <- addHaystackData(proj = projART,
  haystackParentDir = "data/hiv-haystack",
  haystackSamples = haystackSamplePairingART)

projART <- haystackART$newProj
```


```{r}
# proj_clustFilt <- addImputeWeights(proj_clustFilt)
# 
# pImpute <- plotEmbedding(
#     ArchRProj = proj_clustFilt, 
#     colorBy = "GeneScoreMatrix", 
#     name = c("TBX21", "SELL", "CCR7", "IL7R", "IL2RA", "CD27", "CD28", "FAS", "CD19"), 
#     embedding = "UMAP",
#     imputeWeights = getImputeWeights(proj_clustFilt)
# )
# 
# pImputeGrid <- lapply(pImpute, function(x) {
#     gene <- str_extract(x$labels$title, "(\\S+)$")
#     
#     x +
#       guides(color = FALSE, fill = FALSE) + 
#       theme(plot.margin = unit(c(0, 0, 0, 0), "cm")) +
#       labs(title = gene,
#         x = "UMAP Dim 1",
#         y = "UMAP Dim 2") +
#       theme(
#         axis.line = element_line(colour = "black"),
#         axis.text.x = element_blank(),
#         axis.ticks.x = element_blank(),
#         axis.text.y = element_blank(),
#         axis.ticks.y = element_blank(),
#         plot.title = element_text(size = 12),
#         panel.border = element_blank(),
#         panel.background = element_blank(),
#         plot.margin = unit(c(0,0,0,0), "cm"),
#         plot.title.position = "plot"
#       )
# })
# do.call(cowplot::plot_grid, c(list(ncol = 3), pImputeGrid))
```


```{r}
# load processed adt datasets
tsa_catalog <- readRDS("rds/tsa_catalog.rds")
load("rds/invitro_seuratMerged.RData")
load("rds/chronic_seuratMerged.RData")
load("rds/ART_seuratMerged.RData")
```


```{r}
# make final datasets that contains both modalities
inVitroIntersectCbc <- getIntersectBarcodes(adtInVitro, projInVitro)
chronicIntersectCbc <- getIntersectBarcodes(adtChronic, projChronic)
ARTIntersectCbc <- getIntersectBarcodes(adtART, projART)

projInvitro_matched <- projInVitro[inVitroIntersectCbc, ]
projInvitro_matched <- filterClusterByProp(projInvitro_matched, 0.5)
adtInVitro_matched <- subset(adtInVitro, cells = projInvitro_matched$cellNames)

projChronic_matched <- projChronic[chronicIntersectCbc, ]
projChronic_matched <- filterClusterByProp(projChronic_matched, 0.5)
adtChronic_matched <- subset(adtChronic, cells = projChronic_matched$cellNames)

projART_matched <- projART[ARTIntersectCbc, ]
projART_matched <- filterClusterByProp(projART_matched, 0.5)
adtART_matched <- subset(adtART, cells = projART_matched$cellNames)
```


```{r}
# print out initial unlabeled umaps
plotDualUmap(projInvitro_matched, fn = "InVitro_unlabeledCluster_dualUmap")

plotDualUmap(projChronic_matched, fn = "Chronic_unlabeledCluster_dualUmap", embedding = "UMAP2")
plotDualUmap(projChronic_matched, fn = "Chronic_unlabeledCluster_dualUmap_individual",
  embedding = "UMAP2", cluster = "Sample")

plotDualUmap(projART_matched, fn = "ART_unlabeledCluster_dualUmap")
plotDualUmap(projART_matched, fn = "ART_unlabeledCluster_dualUmap_individual", cluster = "Sample")
```


```{r}
# select features that aren't similar to isotype...not sure if this will really be necessary
inVitroIsoSimilar <- returnSimilarToIsotype(tsa_catalog, isoComparisonsInvitro, 0.05, 4)
chronicIsoSimilar <- returnSimilarToIsotype(tsa_catalog, isoComparisonsChronic, 0.05, 4)
ARTIsoSimilar <- returnSimilarToIsotype(tsa_catalog, isoComparisonsART, 0.05, 4)

inVitroAdtToUse <- returnADTNonSimilar(adtInVitro_matched, tsa_catalog, inVitroIsoSimilar)
chronicAdtToUse <- returnADTNonSimilar(adtChronic_matched, tsa_catalog, chronicIsoSimilar)
ARTAdtToUse <- returnADTNonSimilar(adtART_matched, tsa_catalog, ARTIsoSimilar)
```


```{r}
adtInVitro_matched$atacClusters <- projInvitro_matched$Clusters
adtChronic_matched$atacClusters <- projChronic_matched$Clusters
adtART_matched$atacClusters <- projART_matched$Clusters

getBasePanelAndMarkers(seu = adtInVitro_matched, tsa_catalog = tsa_catalog, featuresToUse = inVitroAdtToUse,
  findMarkerMethod = "wilcox",
  pngFn = "outs/png/InVitro_basepanel.png", tsvFn = "outs/tsv/InVitro_initClusterMarkers.tsv")

getBasePanelAndMarkers(seu = adtChronic_matched, tsa_catalog = tsa_catalog, featuresToUse = chronicAdtToUse,
  findMarkerMethod = "wilcox",
  pngFn = "outs/png/chronic_basepanel.png", tsvFn = "outs/tsv/chronic_initClusterMarkers.tsv")

getBasePanelAndMarkers(seu = adtART_matched, tsa_catalog, featuresToUse = ARTAdtToUse,
  findMarkerMethod = "wilcox",
  pngFn = "outs/png/art_basepanel.png", tsvFn = "outs/tsv/art_initClusterMarkers.tsv")
```


```{r}
projInvitro_matched <- assignManualAnnotation(archrProj = projInvitro_matched,
  annotFn = "manualClusterAnnotations/invitro.csv")

adtInVitro_matched$manualClusterAnnot <- projInvitro_matched$manualClusterAnnot
adtInVitro_matched$haystackOut <- projInvitro_matched$haystackOut

plotDualUmap(projInvitro_matched,
  fn = "inVitro_labeledCluster_dualUmap", device = c("png", "svg"),
  cluster = "manualClusterAnnot")

plotDiscreteBar(projInvitro_matched, "inVitro_discreteAbsolute_matched",
  cluster = "manualClusterAnnot",
  graphType = "absolute",
  device = c("png","svg"))

plotDiscreteBar(projInvitro_matched, "inVitro_discreteHivOnly_matched",
  cluster = "manualClusterAnnot",
  device = c("png", "svg"),
  graphType = "hivOnly")
```


```{r}
projChronic_matched <- assignManualAnnotation(archrProj = projChronic_matched,
  annotFn = "manualClusterAnnotations/chronic.csv")

adtChronic_matched$manualClusterAnnot <- projChronic_matched$manualClusterAnnot
adtChronic_matched$haystackOut <- projChronic_matched$haystackOut

plotDualUmap(projChronic_matched,
  fn = "chronic_labeledCluster_dualUmap",
  cluster = "manualClusterAnnot",
  embedding = "UMAP2")

plotDiscreteBar(projChronic_matched, "chronic_discreteAbsolute_matched",
  cluster = "manualClusterAnnot",
  graphType = "absolute")

plotDiscreteBar(projChronic_matched, "chronic_discreteHivOnly_matched",
  cluster = "manualClusterAnnot",
  graphType = "hivOnly")
```


```{r}
projART_matched <- assignManualAnnotation(archrProj = projART_matched,
  annotFn = "manualClusterAnnotations/art.csv")

adtART_matched$manualClusterAnnot <- projART_matched$manualClusterAnnot
adtART_matched$haystackOut <- projART_matched$haystackOut

plotDualUmap(projART_matched,
  fn = "art_labeledCluster_dualUmap",
  cluster = "manualClusterAnnot")

plotDiscreteBar(projART_matched, "art_discreteAbsolute_matched",
  cluster = "manualClusterAnnot",
  graphType = "absolute")

plotDiscreteBar(projART_matched, "art_discreteHivOnly_matched",
  cluster = "manualClusterAnnot",
  graphType = "hivOnly")
```

### BEGIN DIFFERENTIAL ANALYSIS

```{r}
invitro_tcells_check <- !(adtInVitro_matched$manualClusterAnnot %in% c("B", "Monocyte"))
invitro_tcells <- adtInVitro_matched$haystackOut[invitro_tcells_check]
invitro_markers <- findHIVDifferentialMarkers(seu = adtInVitro_matched,
  cellsPos = names(invitro_tcells[invitro_tcells]),
  cellsNeg = names(invitro_tcells[!invitro_tcells]),
  featuresToUse = inVitroAdtToUse,
  tsa_catalog = tsa_catalog)

makeDifferentialLollipop(invitro_markers, "invitro_lollipop")

invitro_activeLaterCells <- adtInVitro_matched$haystackOut[grepl("(Later|Activated)", adtInVitro_matched$manualClusterAnnot) & invitro_tcells_check]
invitro_activeLate_markers <- findHIVDifferentialMarkers(
  seu = adtInVitro_matched,
  tsa_catalog = tsa_catalog,
  cellsPos = names(invitro_activeLaterCells[invitro_activeLaterCells]),
  cellsNeg = names(invitro_activeLaterCells[!invitro_activeLaterCells]),
  featuresToUse = inVitroAdtToUse)

makeDifferentialLollipop(invitro_activeLate_markers, "invitro_lollipop_activeLate")

invitro_earlyCells <- adtInVitro_matched$haystackOut[grepl("(Early|early)", adtInVitro_matched$manualClusterAnnot) & invitro_tcells_check]
invitro_activeLate_markers <- findHIVDifferentialMarkers(
  seu = adtInVitro_matched,
  tsa_catalog = tsa_catalog,
  cellsPos = names(invitro_earlyCells[invitro_earlyCells]),
  cellsNeg = names(invitro_earlyCells[!invitro_earlyCells]),
  featuresToUse = inVitroAdtToUse)

makeDifferentialLollipop(invitro_activeLate_markers, "invitro_lollipop_early")

invitro_renamingDF <- getCellColData(projInvitro_matched, select = c("haystackOut", "manualClusterAnnot")) %>%
  as.data.frame(.) %>%
  mutate(cellNames = projInvitro_matched$cellNames) %>%
  mutate(condensedCluster = case_when(
    grepl("(Early|early)", manualClusterAnnot) ~ "Early",
    grepl("(Later|Activated)", manualClusterAnnot) ~ "LaterAndActivated",
    grepl("Potential Treg", manualClusterAnnot) ~ "Treg",
    TRUE ~ "Other")) %>%
  mutate(condensedHIVCluster = paste0(condensedCluster, "_", haystackOut))

projInvitro_matched <- addCellColData(ArchRProj = projInvitro_matched,
  data = invitro_renamingDF$condensedHIVCluster,
  cells = invitro_renamingDF$cellNames,
  name = "condensedHivCluster")

projInvitro_matched <- addGroupCoverages(ArchRProj = projInvitro_matched,
  groupBy = "condensedHivCluster",
  minReplicates = 3,
  force = TRUE)

projInvitro_matched <- addReproduciblePeakSet(
  ArchRProj = projInvitro_matched,
  groupBy = "condensedHivCluster", 
  pathToMacs2 = "/Users/vince/anaconda3/envs/asapseq/bin/macs2",
  force = TRUE
)

projInvitro_matched <- addPeakMatrix(projInvitro_matched, force = TRUE)

invitro_markerTest_Activated <- getMarkerFeatures(
  ArchRProj = projInvitro_matched, 
  useMatrix = "PeakMatrix",
  groupBy = "condensedHivCluster",
  testMethod = "wilcoxon",
  bias = c("TSSEnrichment", "log10(nFrags)"),
  useGroups = "LaterAndActivated_TRUE",
  bgdGroups = "LaterAndActivated_FALSE",
  maxCells = 1000,
  verbose = FALSE
)

# custom volcano plot (essentially the same as plotMarkers)
activatedVolcano <- data.frame(x = invitro_markerTest_Activated@assays@data@listData$Log2FC$x,
  y = log10(invitro_markerTest_Activated@assays@data@listData$FDR$x) * -1) %>%
  mutate(pointColor = case_when(
    y > -1 * log10(0.05) & x > 0 ~ "Upregulated in HIV+",
    y > -1 * log10(0.05) & x < 0 ~ "Downregulated in HIV+",
    TRUE ~ "Not significant"
)) %>%
  {ggplot(data = ., aes(x = x, y = y, color = pointColor)) +
  geom_point(size = 0.3, alpha = 0.8) +
  theme_classic() +
  labs(x = "log2 Fold Change",
    y = "-log10 FDR",
    color = "") +
  scale_y_continuous(expand = expansion(mult = c(0,0.05)), limits = c(0, NA)) +
  scale_color_manual(values = c(HIVNEGCOLOR, "#bbbbbb", HIVPOSCOLOR))
  } +
  theme(legend.position = "none",
    axis.text = element_text(size = 8),
    axis.title = element_text(size = 8),
    legend.text = element_text(size = 8),
    aspect.ratio = 1) +
  guides(colour = guide_legend(override.aes = list(size = 3), nrow = 3))

ggsave(filename = "outs/svg/invitro_activatedHIVvolcano.pdf",
  plot = activatedVolcano, dpi = "retina", height = 2, width = 2)
rm(activatedVolcano)

projInvitro_matched <- addMotifAnnotations(ArchRProj = projInvitro_matched, motifSet = "cisbp", name = "Motif")

motifsUp <- peakAnnoEnrichment(
  seMarker = invitro_markerTest_Activated,
  ArchRProj = projInvitro_matched,
  peakAnnotation = "Motif",
  cutOff = "FDR < 0.05 & Log2FC > 0")

motifsDown <- peakAnnoEnrichment(
  seMarker = invitro_markerTest_Activated,
  ArchRProj = projInvitro_matched,
  peakAnnotation = "Motif",
  cutOff = "FDR < 0.05 & Log2FC < 0")

plotMotifRank <- function(motifChanged, nLabel = 10) {
  df <- data.frame(TF = rownames(motifChanged), mlog10Padj = assay(motifChanged)[,1])
  df <- df[order(df$mlog10Padj, decreasing = TRUE),]
  df$rank <- seq_len(nrow(df))
  
  g <- ggplot(df, aes(rank, mlog10Padj)) + 
    geom_hline(yintercept = -log10(0.05), color = "#333333", linetype = "dotted") +
    geom_point(size = 1, color = "#00000080", shape = 16) +
    ggrepel::geom_label_repel(
      data = df[rev(seq_len(nLabel)), ], aes(x = rank, y = mlog10Padj, label = TF), 
      size = 6 / ggplot2:::.pt,
      max.overlaps = 30,
      force_pull = 5,
      force = 3,
      segment.color = "#22222280",
      label.padding = 0.16,
      label.size = 0.2,
      color = "black",
      min.segment.length = 0.1
    ) +
    theme_classic() +
    theme(axis.title = element_text(size = 8),
      axis.text = element_text(size = 8)) +
    labs(x = "Rank Sorted TFs Enriched",
      y = "-log10(p adj)")
  
  return(g)
}

ggsave("outs/svg/invitro_motifsUp_activatedHIVneg.svg", plot = plotMotifRank(motifsDown), width = 2.5, height = 2.5, dpi = "retina")
ggsave("outs/svg/invitro_motifsUp_activatedHIVpos.svg", plot = plotMotifRank(motifsUp), width = 2.5, height = 2.5, dpi = "retina")
```


```{r}
projInvitro_matched <- addArchRAnnotations(ArchRProj = projInvitro_matched, collection = "EncodeTFBS")

motifsUpEncode <- peakAnnoEnrichment(
  seMarker = invitro_markerTest_Activated,
  ArchRProj = projInvitro_matched,
  peakAnnotation = "EncodeTFBS",
  cutOff = "FDR < 0.05 & Log2FC > 0")

ggsave("outs/png/invitro_motifsUpEncode_activatedHIVpos.png", plot = plotMotifRank(motifsUpEncode), width = 3, height = 3, dpi = "retina")
```


```{r}
chronic_tcells_check <- !(grepl("(B|APC|CD8)", adtChronic_matched$manualClusterAnnot))
chronic_tcells <- adtChronic_matched$haystackOut[chronic_tcells_check]
chronic_markers <- findHIVDifferentialMarkers(seu = adtChronic_matched,
  cellsPos = names(chronic_tcells[chronic_tcells]),
  cellsNeg = names(chronic_tcells[!chronic_tcells]),
  featuresToUse = chronicAdtToUse,
  findMarkerMethod = "DESeq2",
  tsa_catalog = tsa_catalog)

makeDifferentialLollipop(chronic_markers, "chronic_lollipop")

chronic_renamingDF <- getCellColData(projChronic_matched, select = c("haystackOut", "manualClusterAnnot")) %>%
  as.data.frame(.) %>%
  mutate(cellNames = projChronic_matched$cellNames) %>%
  mutate(condensedCluster = case_when(
    grepl("(Early|Mid|Late|Tfh)", manualClusterAnnot) ~ "T cell",
    TRUE ~ "Other")) %>%
  mutate(condensedHIVCluster = paste0(condensedCluster, "_", haystackOut))

projChronic_matched <- addCellColData(ArchRProj = projChronic_matched,
  data = chronic_renamingDF$condensedHIVCluster,
  cells = chronic_renamingDF$cellNames,
  name = "condensedHivCluster",
  force = TRUE)

projChronic_matched <- addGroupCoverages(ArchRProj = projChronic_matched,
  groupBy = "condensedHivCluster",
  minReplicates = 3,
  force = TRUE)

projChronic_matched <- addReproduciblePeakSet(
  ArchRProj = projChronic_matched,
  groupBy = "condensedHivCluster", 
  pathToMacs2 = "/Users/vince/anaconda3/envs/asapseq/bin/macs2",
  force = TRUE
)

projChronic_matched <- addPeakMatrix(projChronic_matched, force = TRUE)

chronic_markerTest <- getMarkerFeatures(
  ArchRProj = projChronic_matched, 
  useMatrix = "PeakMatrix",
  groupBy = "condensedHivCluster",
  testMethod = "wilcoxon",
  bias = c("TSSEnrichment", "log10(nFrags)"),
  useGroups = "T cell_TRUE",
  bgdGroups = "T cell_FALSE",
  maxCells = 1000,
  verbose = FALSE
)

chronic_activatedVolcano <- data.frame(x = chronic_markerTest@assays@data@listData$Log2FC$x,
  y = log10(chronic_markerTest@assays@data@listData$FDR$x) * -1) %>%
  mutate(pointColor = case_when(
    y > -1 * log10(0.05) & x > 0 ~ "Upregulated in HIV+",
    y > -1 * log10(0.05) & x < 0 ~ "Downregulated in HIV+",
    TRUE ~ "Not significant"
)) %>%
  {ggplot(data = ., aes(x = x, y = y, color = pointColor)) +
  geom_point(size = 0.5) +
  theme_classic() +
  labs(x = "log2 Fold Change",
    y = "-log10 FDR",
    color = "") +
  scale_y_continuous(expand = expansion(mult = c(0,0.05)), limits = c(0, NA)) +
  scale_color_manual(values = c("blue", "grey", "red"))
  } +
  theme(legend.position = "bottom",
    aspect.ratio = 1) +
  guides(colour = guide_legend(override.aes = list(size = 5), nrow = 2))

ggsave(filename = "outs/png/chronic_HIVvolcano.png", plot = chronic_activatedVolcano, dpi = "retina", height = 4.5, width = 4.5)
rm(chronic_activatedVolcano)
```

```{r}
art_tcells_check <- !(grepl("Other", adtART_matched$manualClusterAnnot))
art_tcells <- adtART_matched$haystackOut[art_tcells_check]
art_markers <- findHIVDifferentialMarkers(seu = adtART_matched,
  cellsPos = names(art_tcells[art_tcells]),
  cellsNeg = names(art_tcells[!art_tcells]),
  featuresToUse = ARTAdtToUse,
  findMarkerMethod = "DESeq2",
  tsa_catalog = tsa_catalog)

makeDifferentialLollipop(art_markers, "outs/png/art_lollipop.png")
```

```{r}
viralFragGraphTheme <- list(
  scale_x_continuous(expand = c(0, 0), limits = c(0,NA)),
  theme_classic(),
  theme(
    axis.text.y = element_blank(),
    axis.ticks.y = element_blank(),
    axis.title.y = element_blank(),
    panel.grid.major.y = element_line(color = "#DDDDDD80"),
    strip.background = element_rect(color = "#ffffff", fill = "#ffffff"),
    strip.text = element_text(size = 10)),
  labs(x = "Read fragment aligned to provirus (bp)")
)

haystackInVitro$viralFrags %>%
  ggplot() +
  geom_segment(aes(x = startBp, xend = endBp, y = newCbc, yend = newCbc), color = "#FF000050") +
  viralFragGraphTheme


haystackChronic$viralFrags %>%
  mutate(individual = ifelse(grepl("C01", sample), "C01", "C02")) %>%
  ggplot() +
  geom_segment(aes(x = startBp, xend = endBp, y = newCbc, yend = newCbc), color = "#FF000050", size = 2) +
  facet_grid(individual ~ ., scales = "free_y", space = "free_y") +
  viralFragGraphTheme

```

